microsoft-edge/progressive-web-apps-chromium/how-to/widgets.md

```diff
@@ -0,0 +1,389 @@
+
+```javascript
+self.addEventListener('widgetclick', event => {
+  const action = event.action;
+
+  switch (action) {
+    case "create-meeting":
+      // This is a Widget action event.
+      break;
+    case "widget-save":
+      // A Widget setting was changed.
+      break;
+  }
+});
+```
+
+As shown in the above code snippet, the `widget-save` event is triggered in the service worker when a particular Widget instance has settings and one of the settings was changed.
+
+See the [Service Worker API reference](#service-worker-api-reference) below for more information about the `widgetclick` event and what information you can access from it.
+
+
+<!-- ====================================================================== -->
+## Access Widget instances at runtime
+
+Widget instances can be accessed at runtime from your service worker code.  For example, this approach can can be useful for updating Widgets periodically.
+
+**TODO: Intro/tutorial.**
+
+
+<!-- ====================================================================== -->
+## Service Worker API reference
+
+**TODO: Widgets are too early in the making for MDN. But there is enough complexity with the APIs that a reference doc is needed. The plan is to document all functions and objects right here in this section. Ultimately this would move to MDN and we'll remove this section.**
+
+The service worker global object (or [ServiceWorkerGlobalScope](https://developer.mozilla.org/docs/Web/API/ServiceWorkerGlobalScope)) contains a `widgets` attribute that exposes the following Promise-based methods:
+
+| Method | Description | Parameters | Return value |
+|:--- |:--- |:--- |:--- |
+| `getByTag(tag)` | Get a Widget by tag | The Widget tag | A Promise that resolves to the [Widget object](#widget-object) that matches the tag, or `undefined`. |
+| `getByInstanceId(id)` | Get a Widget by instance ID | The Widget instance ID | A Promise that resolves to the corresponding [Widget object](#widget-object), or undefined.
+| `getByHostId(id)` | Get Widgets by host ID  | The host ID | An array of [Widget objects](#widget-object) found in that host.
+| `matchAll(options)` | Get Widgets by matching options | A [WidgetOptions object](#widgetoptions-object) | A Promise that resolves to an array of [Widget objects](#widget-object) that match the `options` criteria. |
+| `updateByInstanceId(id, payload)` | Update a Widget by instance ID  | The instance ID, and a [WidgetPayload object](#widgetpayload-object) | A Promise that resolves to `undefined` or `Error`.
+| `updateByTag(tag, payload)` | Update a Widget by tag | The Widget tag, and a [WidgetPayload object](#widgetpayload-object) | A Promise that resolves to undefined or Error.
+| `removeByInstanceId(id)` | Remove a Widget by instance ID | The instance ID | A Promise that resolves to undefined or Error.
+| `removeByTag()` | Remove a Widget by tag | The Widget tag | A Promise that resolves to undefined or Error.
```

Comment:
`removeBy*` methods are not supported yet, and may not ever be implemented. 

The Widget Dashboard doesn't support apps removing widgets, and it's unclear if any other Widget hosts will want to allow the app to remove widgets.